server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -0,0 +1,119 @@
+     * updated even if the ClusterState is updated on this node or the remote node. The history is retrieved asynchronously, and only if
+     * requestRemoteMasterHistory has been called for this node. If anything has gone wrong fetching it, or if it has not been fetched in
+     * time for the current request, the returned value will be null.
+     * @param node The node whose MasterHistory we want
+     * @return The MasterHistory from the remote node's point of view. This MasterHistory object will not be updated with future changes
+     */
+    @Nullable
+    public List<DiscoveryNode> getRemoteMasterHistory(DiscoveryNode node) {
+        return nodeToHistoryMap.get(node);
+    }
+
+    /**
+     * This method attempts to fetch the master history from the requested node. If we are able to successfully fetch it, it will be
+     * available in a later call to getRemoteMasterHistory. The client is not notified if or when the remote history is successfully
+     * retrieved. This method only fetches the remote master history once, and it is never updated unless this method is called again.
+     * This is a remote call, so clients should avoid calling it any more often than necessary.
+     * @param node The node whose view of the master history we want to fetch
+     */
+    public void requestRemoteMasterHistory(DiscoveryNode node) {
+        long startTime = System.nanoTime();
+        transportService.openConnection(
+            // Note: This connection is explicitly closed onResponse or onFailure of the request below
+            node,
+            ConnectionProfile.buildDefaultConnectionProfile(clusterService.getSettings()),
+            new ActionListener<>() {
+                @Override
+                public void onResponse(Transport.Connection connection) {
+                    logger.trace("Opened connection to {}, making master history request", node);
+                    transportService.sendRequest(
+                        node,
+                        MasterHistoryAction.NAME,
+                        new MasterHistoryAction.Request(),
+                        new ActionListenerResponseHandler<>(new ActionListener<>() {
+
+                            @Override
+                            public void onResponse(MasterHistoryAction.Response response) {
+                                connection.close();
+                                long endTime = System.nanoTime();
+                                logger.trace("Received history from {} in {}", node, TimeValue.timeValueNanos(endTime - startTime));
+                                nodeToHistoryMap.put(node, response.getMasterHistory());
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                connection.close();
+                                logger.error("Exception in master history request to master node", e);
```

Comment:
I think `error` is not the right level - see https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#error for some guidance. I'm not even sure `warn` logging is necessary, we will already be logging all sorts of other stuff if we get to this point, although I think it'll be ok and we can revisit it later if it's too noisy.

However, we do need to handle the exception more meaningfully. The design docs call failures here "notable", i.e. we would want to see them in the health API response.

We should also send this request with a timeout - taking an unreasonably long time to respond would also be a notable failure. 10s seems like a reasonable default timeout.

Reply:
This is happening asynchronously, so there's not a way to put it in any response is there? Best case we could put it on the response to whatever request comes in next... but then we'd wind up with the confusion we're seeing on the deprecation logger for that kind of thing right?

Reply:
I mean we should capture the failure so we can return it in the health API. "We think the master is flaky and when we asked the master for its status it blew up with the following exception" is a useful thing to expose.

Reply:
Let me make sure I understand what you're saying:

- Whenever we get an exception in either of these two places, store that exception off somewhere (e.g. a list in MasterHistoryService).
- On every stable master health check request, look in that location as the first step, and if there's something there:
  - Immediately return a YELLOW(?) response with any exceptions there
  - Clear out the exceptions so that the next request doesn't repeat them

Reply:
Assuming the above is true:
Do we want to guarantee that no exception is returned in more than one response? I assume yes.
Do we want to guarantee that an exception is not lost? That is, if the response fails for some reason, do we want to "roll back" taking it out of the list of exceptions? I'm assuming that that kind of failure is rare enough and this information is not quite important enough (since it's also in the logs and not totally lost) to warrant that kind of complexity.
Do we want to attempt to have all exceptions (if there are more than one) come back in a single response? That would mean a little more locking around the list of exceptions, but maybe a little less confusion for users (or maybe a little more, I'm not sure). I'm on the fence about this one.

Reply:
I don't think we should strive to return the exception only once. My mental model here is that our view of the remote master history isn't always a `List<DiscoveryNode>`, it might also be an exception. If it's an exception then we can't make judgements about whether the remote master thinks it's stable or not, but that's ok, we can still report the exception.

Reply:
In a sense `MasterHistoryService#nodeToHistoryMap` is a cache of (our view of) the remote history on various nodes. On reflection perhaps we should just keep one remote history, rather than one per node. It doesn't really matter which node's history we check, we just need a second opinion to try and rule out local issues. Technically the current implementation is a (very slow) leak since we never prune entries for old masters.

Reply:
That's a good point (that it doesn't matter which remote history we get). Is there not any edge case where we'd regret using the history of one master vs another? I definitely don't have the experience to know for sure. I'll go ahead and make this change though b/c that makes sense, and would simplify things a little.

Reply:
> Is there not any edge case where we'd regret using the history of one master vs another?

Maybe, but this is all very approximate anyway and I don't think we'll see much difference either way. Let's take the simpler path here at least until we've some empirical evidence that it needs something cleverer.